### DISCUSSION TITLE
Feature Requests Triage at the Verus Retreat '24

### DISCUSSION THREAD

**Discussion by utaal** [verus-lang/verus#1176] [Feature requests]
Add a feature request, up vote it if you care about it, and add comments on why you want it.

**Use `## Feature title` as the comment header.**

**Comment by utaal** [comment]
## Lightweight [termination checking for exec](https://github.com/verus-lang/verus/discussions/254)

Ignore liveness.

**Reply by utaal** [reply]
More sources of non-termination than just checking for decreases clauses on loop and recursion.

**Reply by utaal** [reply]
Already supported for recursion, and loops. No enforcement yet.

How to do UX?
error (initially warning) in the absence of a decreases clause in `exec` code, and crate-level attributes to opt-in or opt-out of total correctness.

**Reply by utaal** [reply]
@matthias-brun, @jaylorch: consider opting out for individual function, i.e. `main`, with an attributes. Folks are not against the idea.

@jaylorch: find a name flaggable by the auditor.

**Comment by utaal** [comment]
## [From and Into traits](https://github.com/verus-lang/verus/discussions/1129#discussioncomment-9492707)

**Reply by utaal** [reply]
You can do it for whatever type you want https://doc.rust-lang.org/std/convert/trait.From.html

`?` desugars into something that has a trait bound that interacts with `FromResidual`.

* [ ] proper spec for `From` and remove hack for `?`.

**Reply by utaal** [reply]
One can add their own external spec.

* [ ] document that process.

**Comment by utaal** [comment]
## [Vector literals](https://github.com/verus-lang/verus/discussions/1129#discussioncomment-9492710)

**Reply by utaal** [reply]
@tjhance, @Chris-Hawblitzel: it may be pretty easy to support.

**Comment by utaal** [comment]
## vstd functions for [(debug) printing](https://github.com/verus-lang/verus/discussions/1129#discussioncomment-9736972)

**Reply by utaal** [reply]
@jaybosamiya: something that's easier than `println`

It's supposed to be transparent. Encode the body, ignore the `dbg!`.

Add `Debug` to the ignored traits.

**Reply by utaal** [reply]
@Chris-Hawblitzel: find all the things we want Verus to ignore.

**Reply by jaybosamiya-ms** [reply]
Specifically, `dbg!` (https://doc.rust-lang.org/std/macro.dbg.html)

**Reply by utaal** [reply]
@Chris-Hawblitzel treat the `Debug` impl as external, and enforce it to be called from the `dbg!` macro.

Add an attribute "this comes from the `dbg!` macro".

**Comment by utaal** [comment]
## Support for [dyn](https://github.com/verus-lang/verus/discussions/1047)

**Reply by marshtompsxd** [reply]
A use case that also motivates this: in the Anvil project we have to define a struct that needs to take a sequence of different types that all implement a trait T. Right now our solution is: (1) define an opaque type and use that type in the struct, (2) define (uninterpretable) functions to convert between each type to/from the opaque type, (3) define axioms for each type that any object of this type remains unchanged after a RTT of the conversion --- similar to marshal and unmarshal.

**Comment by utaal** [comment]
## [Type invariants](https://github.com/verus-lang/verus/discussions/962)

**Reply by utaal** [reply]
High priority feature for @tjhance. But what do people want?
* do you want invariants to be temporarily broken?

@y1ca1: don't want the invariant to ever be broken, and in the use case it's not going to be mutable.

**Reply by utaal** [reply]
James G.: have two different types, one with the invariant enforced, and one without.

**Comment by utaal** [comment]
## Separating and identifying [trusted vs. untrusted code](https://github.com/verus-lang/verus/discussions/112)

**Reply by utaal** [reply]
Discussed in the [discussion](https://github.com/verus-lang/verus/discussions/112).

**Comment by utaal** [comment]
## [Dual mode spec/exec](https://github.com/verus-lang/verus/discussions/851)

**Comment by utaal** [comment]
## ['returns' clause for exec functions](https://github.com/verus-lang/verus/discussions/851)

**Reply by utaal** [reply]
returns clause one step forward;
maybe other proposals are two steps forward.

**Comment by utaal** [comment]
## [make eager checking for spec fn requires configurable](https://github.com/verus-lang/verus/discussions/764#discussioncomment-7188162)

**Comment by utaal** [comment]
## Add support for `seq![0; 10]` (repeat)

It's easy to do, and ChatGPT invented it!

**Reply by jaybosamiya-ms** [reply]
Solved by https://github.com/verus-lang/verus/pull/1178

**Comment by utaal** [comment]
## Support for some fields pub and some private in a struct

@utaal: it's relatively easy.

**Comment by utaal** [comment]
## Smoke test

Insert `assert(false)`, or perhaps focus on `assert(precondition => false)`.

## Using `(unsat-core)` to see which conjuncts were actually used for a proof

Also potentially run the query multiple times.

Oded: Environment with precondition with 10 conjuncts, and only 2 are needed.

**Comment by utaal** [comment]
## Document `external_*`

* `external_fn_specification`
* `external_type_specification`
* `external_trait_specification`

Not many votes, but asked verbally many times.

**Comment by utaal** [comment]
These have been folded into the relative discussions and otherwise ported over to https://github.com/orgs/verus-lang/projects/1
